### ISSUE TITLE
A recursively saved FMModel fails to re-open as a new FMModel instance

### ISSUE THREAD

**Issue by LisaWeijers** [Deltares/HYDROLIB-core#211]
**Describe the bug**
I opened an existing model with mdu.model --> FMModel. I saved it recursively to a new folder. When I open the model I just saved again, I get a lot of errors. 

**To Reproduce**
Using the moergestels_broek model (because I'm using for everything I try everything nowadays)
```
from pathlib import Path
from hydrolib.core.io.mdu.models import FMModel

model_folder = Path("playground")
old_mdu = model_folder/'model'/'moergestels_broek.mdu'
new_mdu= model_folder/'iterations'/'new.mdu'

old_mdu = FMModel(old_mdu)
old_mdu.save(new_mdu, recurse=True)

mdu = FMModel(new_mdu)
print(mdu.show_tree())
```

Result (snippets): 
```
geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 152 -> prof_RS1_RVK-DP148 -> yCoordinates -> 0
  value is not a valid float (type=type_error.float)
geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 152 -> prof_RS1_RVK-DP148 -> zCoordinates -> 0
  value is not a valid float (type=type_error.float)
geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 152 -> prof_RS1_RVK-DP148 -> frictionPositions -> 0
  value is not a valid float (type=type_error.float)
...
original.mdu -> geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 176 -> prof_RS376.1-DP3 -> yCoordinates -> 0
  value is not a valid float (type=type_error.float)
original.mdu -> geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 176 -> prof_RS376.1-DP3 -> zCoordinates -> 0
  value is not a valid float (type=type_error.float)
original.mdu -> geometry -> crossDefFile -> cross_section_definitions.ini -> definition -> 176 -> prof_RS376.1-DP3 -> frictionPositions -> 0
  value is not a valid float (type=type_error.float)
...
original.mdu -> output -> mapInterval -> 0
  value is not a valid float (type=type_error.float)
original.mdu -> output -> rstInterval -> 0
  value is not a valid float (type=type_error.float)
original.mdu -> output -> waqInterval -> 0
  value is not a valid float (type=type_error.float)
python-BaseException

```

**Expected behavior**
I expected that I can simply read `new.mdu` using `FMModel`
I attached the original MDU and cross_section_definitions.ini & the copied ones (called something with new)

[cross_section_definitions_new.ini.txt](https://github.com/Deltares/HYDROLIB-core/files/7979642/cross_section_definitions_new.ini.txt)
[moergestels_broek.mdu.txt](https://github.com/Deltares/HYDROLIB-core/files/7979638/moergestels_broek.mdu.txt)

[cross_section_definitions_new.ini.txt](https://github.com/Deltares/HYDROLIB-core/files/7979635/cross_section_definitions_new.ini.txt) (simply called 'cross_section_definitions.ini', as referenced in the mdu)
[new.mdu.txt](https://github.com/Deltares/HYDROLIB-core/files/7979614/new.mdu.txt)

**Version info (please complete the following information):**

OS: Windows 10
Version: 0.2.0
Python: 3.8

**Comment by LisaWeijers** [comment]
Additionally, I tried to open the MDU of the model that I created (so after reading moergestels_broek.mdu, changing nothing, and saving the MDU with `mdu.save('original.mdu', recurse=True`) in DHydro and it also won't even open there. 
(please bare with me that I am naming this new file 'original.mdu', it makes sense with the context of the profile optimizer, I promise :P)

![Error messages in DHydro](https://user-images.githubusercontent.com/28150338/152530518-28fcf8d8-b6ae-485e-904c-40b628836451.png)

A few things that I notice here:

- It says it expects some files based on the MDU, which haven't been copied with `original.save('name.mdu', recurse=True)`, namely obspts_obs.ini & structures.ldb. 
- The compound structure that is included in structure_temp.ini is exported with an empty fields (third message in screenshot). It's about the line I marked in the screenshot below
![Compound structure snippet](https://user-images.githubusercontent.com/28150338/152531684-8cff36f2-a188-4a7d-ae80-5d8371392cba.png)
- I can't find there the "Value '0.0;0.0;0.0' in collection cannot be read" etc exactly, but I noticed this format in crosssec_def.ini files for sure. In the source file the format for lists of numbers is "0.0 0.0 0.0" etc (so space as seperator) but in the file saved through hydrolib it becomes "0.0;0.0;0.0" etc (see screenshot below)
![Format of lists in crosssec_def](https://user-images.githubusercontent.com/28150338/152532672-e165df77-f4b1-4483-ba01-4c086f660690.png)

When I fix the first two bullet points (manually copying  obspts_obs.ini & structures.ldb to a new folder & deleting the lines in structure_temp.ini, the next error it gives is about the crosssec_def number format: 
"0.0;1.451;3.996;4.997;6.704;8.461;8.58;9.018;9.056;12.48;16.063;16.107;16.59;16.652;18.3;20.173;21.717;25.186 is not a valid value for Double."
 
**In summary**
- not all files defined in the MDU will be saved with MDU.save(new_path/name.mdu, recurse=True)
- lists of numbers are saved with wrong seperators

**Comment by arthurvd** [comment]
I investigated what causes this wrong errors, and below is some input for a technical discussion:

**Cause:**
Early in the serialization process, any INIBasedModel will go through the static method `_convert_value()`, which joins list values with a hardcoded semicolon:
```python
            return ";".join([str(x) for x in v])
```

**Start of a solution:**
I could change this into a `@classmethod`, and then use the class's already existing `_list_delimiter()` class method, as follows:
```python
            return cls.list_delimiter().join([str(x) for x in v])
```
This solved the reported issue for some structure files.
However, that (sub)class level is still not detailed/deep enough: within a single class, some of its fields might have a different list value delimiter than others.
For example:
```python
class ZWRiverCrsDef(CrossSectionDefinition):
#[..]
    _split_to_list = get_split_string_on_delimiter_validator(
        "levels",
        "flowwidths",
        "totalwidths",
        "frictionvalues",
        delimiter=" ",
    )
    _split_to_list2 = get_split_string_on_delimiter_validator(
        "frictionids",
        "frictiontypes",
        delimiter=";",
    )
```

**Question:**
How can we nicely define list value delimiters at the field level, such that both the parser (post-validator) and the serializer make use of that same setting/class field?

**Comment by arthurvd** [comment]
@BeardedPlatypus wrote Yesterday at 7:26 PM:
> I think the easiest approach would be to define a default like you did in your answer, and an additional delimiter dictionary. The delimiter dictionary will be empty by default, and will let us overwrite the delimiter of specific fields. We can then use the field name to actually retrieve the delimiter. At least that would be my first idea

@evetion wrote Yesterday at 8:52 PM
> Wel agreed, you could also store it as `should_be_forbidden_field: List[float] = Field(..., delimiter=";")`
And then create a helper method on our BaseClass to access custom Field properties

**Comment by LisaWeijers** [comment]
Alright, it took a while to get back to this because I'm both trying to test if this is robust & I'm working for the profile optimizer at the same time. But I think I figured some things out about some thing

I used this version:

- lists of numbers are saved with wrong seperators
--> probably fixed in PR https://github.com/Deltares/HYDROLIB-core/pull/212.

This works better! I can save a copy of an existing FM model, and the copy can be read by Hydrolib again. So partially solved. 

My intention (which is in hindsight, not the best approach, I found an neater alternative) was to do this:
BASE MODEL --> change a profile --> ITERATION 1 --> change a profile --> ITERATION 2 --> change a profile --> etc
This produced a new error. Here is a base code example that produces the bug:
```
old_mdu_path = r'c:\dhydro\20211102_Developers_tutorial_Hydrolib\code\data\model\moergestels_broek.mdu'
model_folder = Path("playground")

old_mdu = FMModel(Path(old_mdu_path))
mdu_1_path = Path(model_folder)/'test_1' / 'save_1.mdu'
old_mdu.save(mdu_1_path, recurse=True)

mdu_1 = FMModel(mdu_1_path)
mdu_2_path = Path(model_folder)/'test_2' / 'save_2.mdu'
mdu_1.save(mdu_2_path, recurse=True)

mdu_2 = FMModel(mdu_2_path)
print(mdu_2.show_tree())
```
This crashes at `mdu_1.save(mdu_2_path, recurse=True)` with the following output:

```
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 484, in file_load_context
    yield file_loading_context
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 670, in save
    self._save_tree(context)
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 715, in _save_tree
    save_traverser.traverse(self, context)
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 240, in traverse
    acc = self._post_traverse_func(model, acc)  # type: ignore[arg-type]
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 706, in save_post
    model._save()
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 723, in _save
    self._serialize(self.dict())
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\io\ini\models.py", line 190, in _serialize
    write_ini(self._resolved_filepath, self._to_document(), config=config)
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\io\ini\models.py", line 184, in _to_document
    sections.append(value._to_section())
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\io\ini\models.py", line 121, in _to_section
    prop = Property(
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 52, in __init__
    raise e
  File "C:\ProgramData\Anaconda3\envs\hydrolib_resave\lib\site-packages\hydrolib\core\basemodel.py", line 48, in __init__
    super().__init__(**data)
  File "pydantic\main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Property
comment
  str type expected (type=type_error.str)
```
What it's trying to work with is this:
`{'key': '1d2dLinkFile', 'value': '', 'comment': ['File containing the custom parameterization of 1D-2D links.', 'File containing the custom parameterization of 1D-2D links.']}`
The value for comment is a list, which shouldn't happen, it should be a string. 

I can work around it for the profile optimizer, namely by always doing my changes on a base model, not on the previous iterations. But hopefully this report can help make hydrolib more robust ;) 

@arthurvd 
Shall I close this issue for now, since it is fixed with #212 and make a new issue for the 1d2dlinkfile comment thing? Or do we want to keep this open for further improvements?
